src/resources/SearchAnalysis/SearchAnalysisInterfaces.ts

```diff
@@ -0,0 +1,158 @@
+    | 'AUTHENTICATION_SEARCH_HUB'
+    | 'ML_ROUTING'
+    | 'PARAMETER_PIPELINE'
+    | 'CONDITION_ROUTING'
+    | 'DEFAULT';
+
+export interface ExecutionStep<T extends AvailableExecutionStep> {
+    type: T;
+    name: string;
+}
+export interface Condition {
+    id: string;
+}
+export interface AffectedRequestParameters {
+    affectedRequestParameters: Array<{
+        name: string;
+        value: unknown;
+    }>;
+}
+export interface AffectedResultsPosition {
+    affectedResultsPosition: number[];
+}
+
+export type Applied = {id: string; condition?: Condition};
+
+export interface AppliedRules<Affected = Record<string, unknown>, OtherPropertiesApplied = Record<string, unknown>> {
+    appliedCount: number;
+    applied: Array<Applied & Affected & OtherPropertiesApplied>;
+}
+
+export interface SelectedPipelineDefinitiion {
+    id: string;
+    name: string;
+    condition?: Condition;
+}
+
+export interface PipelineSelectionCauseABTest {
+    originalPipeline: {
+        id: string;
+        name: string;
+        condition?: Condition;
+    };
+    targetPipeline: {
+        id: string;
+        name: string;
+        condition?: Condition;
+    };
+}
+
+export interface PipelineSelectionCause {
+    type: AvailablePipelineSelection;
+    abTest?: PipelineSelectionCauseABTest;
+}
+
+export interface QueryPipelineSelection extends ExecutionStep<'pipelineSelection'> {
+    selectedPipeline: SelectedPipelineDefinitiion;
+    selectionCause: PipelineSelectionCause;
+}
+
+export type QueryParamOverrides = ExecutionStep<'paramOverrides'> & AppliedRules<AffectedRequestParameters>;
+
+export type Thesaurus = ExecutionStep<'thesaurus'> & AppliedRules<AffectedRequestParameters>;
+
+export type StopWords = ExecutionStep<'stopWords'> & AppliedRules<AffectedRequestParameters>;
+
+export type Filters = ExecutionStep<'filters'> & AppliedRules<AffectedRequestParameters>;
+export type RankingExpressions = ExecutionStep<'rankingExpressions'> &
+    AppliedRules<
+        AffectedResultsPosition,
+        {
+            boost: number;
+        }
+    >;
+
+export type FeaturedResults = ExecutionStep<'featuredResults'> & AppliedRules<AffectedResultsPosition>;
+
+export type RankingWeights = ExecutionStep<'rankingWeights'> & AppliedRules;
+
+export type ContentRecommendation = ExecutionStep<'contentRecommendation'> & AppliedRules<AffectedResultsPosition>;
+
+export type ProductRecommendation = ExecutionStep<'productRecommendation'> & AppliedRules<AffectedResultsPosition>;
+
+export interface AutomaticRelevanceTuning extends ExecutionStep<'art'>, AppliedRules<AffectedResultsPosition> {}
+
+export interface DynamicNavigationExperience extends ExecutionStep<'dne'>, AppliedRules<AffectedResultsPosition> {}
+
+export interface Triggers extends ExecutionStep<'triggers'>, AppliedRules {}
+
+export interface IndexQuery extends ExecutionStep<'indexQuery'> {
+    request: Record<string, unknown>;
+}
+
+export type MappedExecutionSteps = {
```

Comment:
Bear with me on this one 😄 

It's so it's easier to do type inference on the caller side, based on the execution step type.

For example, in the admin, I'm able to have a small utility function to do this:

```
export const findByType = <ExecutionStepType extends AvailableExecutionStep>(
    execution: ExecutionSteps,
    executionType: ExecutionStepType
) => {
    const found = execution.find((step) => step.type === executionType);
    return found as MappedExecutionSteps[ExecutionStepType];
};

// properly typed here
const art = findByType(execution, 'art');
```

Reply:
![image](https://user-images.githubusercontent.com/12366410/196235444-ac77cf4a-ad7d-4594-b37b-35fb5bdb7f87.png)
Nice job. It's OK for the `platform-client` to have complex interfaces, it's actually its job to own this responsibility.

Allow me to do one better tho for your e.g.:

```ts
const findByType = <ExecutionStepType extends AvailableExecutionStep, ReturnType extends MappedExecutionSteps[ExecutionStepType]>(
    execution: ExecutionSteps,
    executionType: ExecutionStepType
) => {
    const found = execution.find<ReturnType>(
        (step): step is ReturnType => step.type === executionType
    );
    return found;
};
// Still properly typed here, but no type cast
const art = findByType(execution, 'art');
```

The trick is to provide the type to `Array.find` and type the predicate as a type predicate (`x is Foo`),
this avoids a type-cast.

Reply:
Also possible to do in 'one-line'
```ts
const findByType = <
    ExecutionStepType extends AvailableExecutionStep,
    ReturnType extends MappedExecutionSteps[ExecutionStepType]
>(
    execution: ExecutionSteps,
    executionType: ExecutionStepType
) => execution.find<ReturnType>((step): step is ReturnType => step.type === executionType);
```